shared-bindings/gamepad/GamePad.c

```diff
@@ -96,23 +105,86 @@
 //|
 STATIC mp_obj_t gamepad_make_new(const mp_obj_type_t *type, size_t n_args,
         const mp_obj_t *args, mp_map_t *kw_args) {
-    if (n_args > 8) {
-        mp_raise_TypeError(translate("too many arguments"));
+    if (n_args > 8 || n_args == 0) {
+        mp_raise_TypeError(translate("argument num/types mismatch"));
     }
     for (size_t i = 0; i < n_args; ++i) {
-        if (!MP_OBJ_IS_TYPE(args[i], &digitalio_digitalinout_type)) {
-            mp_raise_TypeError(translate("expected a DigitalInOut"));
-        }
-        digitalio_digitalinout_obj_t *pin = MP_OBJ_TO_PTR(args[i]);
-        raise_error_if_deinited(
-            common_hal_digitalio_digitalinout_deinited(pin));
+        validate_pin(args[i]);
     }
-    if (!MP_STATE_VM(gamepad_singleton)) {
-        gamepad_obj_t* gamepad_singleton = m_new_obj(gamepad_obj_t);
+    gamepad_obj_t* gamepad_singleton = MP_STATE_VM(gamepad_singleton);
+    if (!gamepad_singleton) {
+        gamepad_singleton = m_new_obj(gamepad_obj_t);
         gamepad_singleton->base.type = &gamepad_type;
-        MP_STATE_VM(gamepad_singleton) = gc_make_long_lived(gamepad_singleton);
+        gamepad_singleton = gc_make_long_lived(gamepad_singleton);
+        MP_STATE_VM(gamepad_singleton) = gamepad_singleton;
+    }
+    for (size_t i = 0; i < 8; ++i) {
+        gamepad_singleton->pins[i] = NULL;
+    }
+    gamepad_singleton->pulls = 0;
+    for (size_t i = 0; i < n_args; ++i) {
+        digitalio_digitalinout_obj_t *pin = MP_OBJ_TO_PTR(args[i]);
+        if (common_hal_digitalio_digitalinout_get_direction(pin) !=
+            DIRECTION_INPUT) {
+            common_hal_digitalio_digitalinout_switch_to_input(pin, PULL_UP);
+        }
+        digitalio_pull_t pull = common_hal_digitalio_digitalinout_get_pull(pin);
+        if (pull == PULL_NONE) {
+            common_hal_digitalio_digitalinout_set_pull(pin, PULL_UP);
+        }
+        if (pull != PULL_DOWN) {
+            gamepad_singleton->pulls |= 1 << i;
+        }
+        gamepad_singleton->pins[i] = pin;
     }
-    gamepad_init(n_args, args);
+    return MP_OBJ_FROM_PTR(MP_STATE_VM(gamepad_singleton));
+}
+
+
+//| .. class:: GamePadShift(data, clock, latch)
```

Comment:
Please put this in it's own file for consistency. validate_pin can go in __init__.c as a helper for each.